# <span style="color: black; font-family: Babas Neue; font-size: 3em;">Tutoriel 1.0</span>
<hr height="3" style="color:purple;background-color:purple;"></hr>
<span style="color: black; font-family: Babas Neue; font-size: 1em;">Extraction d'articles depuis les archives de La Presse</span>

<br>
<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
**Bienvenue à notre premier tutoriel!** Nous travaillerons ici sur un script qui nous permettra de parcourir les archives du quotidien La Presse. Nous avons choisi La Presse pour notre tutoriel car ses archives respectent une forme standard, ce qui permet au code d'être facilement repensé pour un autre projet d'extraction de données. <br><br>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Pour mener à bien ce projet, nous utiliserons le langage python à l'intérieur de l'environnement Anaconda. [Anaconda](https://www.continuum.io/downloads) est un écosystème d'applications, de paquets et de modules pour les sciences des données. Il est fort probable que ce que vous recherchez soit contenu dans l'environnement Anaconda de base, mais, au besoin, il est toujours possible d'installer manuellement un module ou une application.
<br><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Si vous souhaitez exécuter les codes par vous-même et explorer, veuillez installer la suite Anaconda pour python 3.6. Sinon, vous pouvez aussi lire ce fichier - sans l'exécuter - comme une sorte de rapport, car j'y ai intégré des commentaires ainsi que les résultats des scripts.
</span></div>

> <div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Pour notre projet, nous utiliserons les modules suivant (venant par défaut avec l'environnement anaconda)<br><br>
- [urllib.request](https://docs.python.org/3/library/urllib.request.html) (Pour envoyer les requêtes URL)<br>
- [BeautifulSoup4](https://www.crummy.com/software/BeautifulSoup/bs4/doc/) (Abbrévié 'bs4'. Sert à traiter le contenu URL)<br>
- [pandas](https://pandas.pydata.org/pandas-docs/stable/10min.html) (Analyse de bases de données en langage python. L'équivalent d'Excel en ligne de commande. Utilisé pour toute les manipulations de bases de données)<br>
- [numpy](http://www.numpy.org/) (Programmation scientifique. Pandas est construit par-dessus l'infrastructure de NumPy, mais ce dernier contient beaucoup de fonctions et algorithmes qui lui sont propres.)
</span></div>

> <div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
De plus, nous utiliserons le module [newspaper](http://newspaper.readthedocs.io/en/latest/), développé en open-source par Lucas Ou-Yang. Celui-ci ne fait pas parti de l'écosystème Anaconda, nous devrons donc l'installer manuellement. Nous pourrions créer un script de *data mining* sans utiliser le module newspaper, mais celui-ci va nous sauver du temps et alléger le processus de codage. <br><br>

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Voici les grandes étapes de notre extraction :<br><br>1) Générer une liste de dates. Dans notre cas, il s'agit des jours entre le 1er décembre 2012 et la dernière journée de juin 2013. <br>2) Générer, par une boucle, une liste de liens vers les archives. Pour se faire, nous partirons de la racine des archives et nous allons ajouter les extensions pour chaque journée. <br> 3) Générer, à partir de cette liste, une autre liste contenant, elle, liens vers les articles. <br> 4) Interroger les articles et en extraire le contenu lorsque désiré. C'est ici que nous utiliserons le module newspaper. <br><br><br>***Afficher la version de la suite anaconda***</span></div>

###### <span style="color: black; font-family: Babas Neue; font-size: 2em;">0.0 Exemple</span>

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Ce document contient des zones de texte (comme ici), des zones de script (vous pouvez les voir avec une invite de commande 'In [x]') situé sur la gauche. La réponse (*output*) du script figure en bas des zones de script.<br><br><br>***Afficher la version de la suite anaconda*** (ZONE DE TEXTE)</span></div>

In [7]:
!conda --version # ZONE DE SCRIPT

conda 4.3.23


<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"><br>Dans la zone de script, les mots suivis par un # sont des commentaires pour vous aider à comprendre. Ils n'ont aucun effet sur votre code<br><br>*** Afficher le résultat d'une équation***</span></div>

In [8]:
x = 10                        # Créer une variable 'x' avec le nombre entier 10 comme valeur
equation_ex = x*(4+1/(2*3)-2) # Les opérateurs mathématiques sont semblables (* = multi, / = div)
print (equation_ex)           # Afficher la variable 'equation_ex', dont la valeur est le résultat de l'équation

21.66666666666667


###### <span style="color: black; font-family: Babas Neue; font-size: 2em;">1. Importation</span>

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Pour commencer, nous allons importer les modules que nous avons déjà dans l'environnement Anaconda, puis installer le module newspaper.<br><br><br>***1.1 Importation des modules***</span></div>

In [9]:
import urllib.request
import bs4
import numpy as np  # 'as' signifie que que nous allons abbrévier le nom, puisque ce sont des modules souvent utilisés
import pandas as pd # Avant : pandas.DataFrame(...), Maintenant : pd.DataFrame(...)

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"><br>*** 1.2.1 Installer newspaper (exécuter dans un shell ou terminal) ***</span></div>


In [10]:
!pip install newspaper3k

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Comme j'ai déjà installé le module manuellement, la machine me répond que les exigences sont déjà satisfaites.

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"><br>*** 1.2.2 Importer newspaper***</span></div>


In [11]:
import newspaper

###### <span style="color: black; font-family: Babas Neue; font-size: 2em;">2. Création d'une liste de date</span>

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Nous allons  générer une liste des dates qui couvrent notre étude. Nous le faisons pour plusieurs raisons, notamment parce que les addresses URL de La Presse intègrent les dates et parce que, de toute façon, nous allons entrer la donnée 'DATE' dans la base de données que nous allons créer. Elle sera déjà faite. <br><br><br>***2.1 Génération d'une liste de date avec numpy (np) et pandas (pd)***</span></div>

In [12]:
array_of_time = pd.to_datetime(np.arange('2012-12-01', '2013-07-01', dtype='datetime64[D]'))
print (array_of_time)

DatetimeIndex(['2012-12-01', '2012-12-02', '2012-12-03', '2012-12-04',
               '2012-12-05', '2012-12-06', '2012-12-07', '2012-12-08',
               '2012-12-09', '2012-12-10',
               ...
               '2013-06-21', '2013-06-22', '2013-06-23', '2013-06-24',
               '2013-06-25', '2013-06-26', '2013-06-27', '2013-06-28',
               '2013-06-29', '2013-06-30'],
              dtype='datetime64[ns]', length=212, freq=None)


<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Les '...' que vous voyez dans la réponse signifient que vous ne voyez pas l'entièreté de la liste. On peut constater, à cette fin, que la liste contient 212 entrées, ce qui est exact si l'on souhaite lister les jours du 1er décembre 2012 à la dernière journée de juin 2013 inclusivement. <br><br> Non seulement nous avons dressé rapidement une liste, mais nous avons également formatté les données pour qu'elles soient perçues comme un objet de type 'date' plutôt que comme une chaîne de caractères normale, (Comme 'Roger'). Ces données contiennent donc maintenant des attributs propres aux dates.

In [13]:
print (array_of_time[0]) # Imprimer la première donnée de notre liste de date (liste[index])
print ('2012-12-01 00:00:00') # Imprimer une chaîne de caractères standard qui semble identique

2012-12-01 00:00:00
2012-12-01 00:00:00


In [14]:
print (array_of_time[0].year) # Imprimer des attributs de classe 'DATE' avec notre variable formaté
print (array_of_time[0].day) 
print (array_of_time[0].month) 
print (array_of_time[0].weekday_name)  
print ('.. Il existe beaucoup plus de fonctions! ..')

2012
1
12
Saturday
.. Il existe beaucoup plus de fonctions! ..


In [15]:
print ('2012-12-01 00:00:00'.year) # Essayer de faire une même fonction avec la chaîne de caractère

AttributeError: 'str' object has no attribute 'year'

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Nous pouvons constater ici qu'une chaîne de caractères standard ne possède pas l'attribut 'année', puisqu'elle donne une réponse d'erreur. Notre première donnée de la liste est convertie comme un objet de type 'date', ce qui la permet de posséder cet attribut. Vous voyez donc l'intérêt de *formater* les données en plus de les générer à grande échelle. 

###### <span style="color: black; font-family: Babas Neue; font-size: 2em;">3. Générer une liste de liens URL vers les archives</span>

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Le système d'archives de La Presse fonctionne ainsi :
<br><br>
http://www.lapresse.ca/ archives / années / mois / jour . php
<br><br>
'archives' est une chaîne de caractères (il ne changera pas).
'années', 'mois' et 'jour' sont les valeurs de nos variables de dates (ils changeront en fonction de nos dates). 
'.php' est le type de fichier hébergé sur le site. Les plus populaires sont .php et .html.  
<br><br>
À partir de notre première liste de dates, nous allons générer une liste de liens vers les archives. Il ne s'agit pas de notre cible final, c'est-à-dire qu'il faudra ensuite aller à l'intérieur de ses liens pour produire une troisième liste, qui est constituée des adresses vers les articles.

<br><br><br>***3.1 Définition de la souche*** (Ce qui ne change pas)</span></div>

In [16]:
root_url = 'http://www.lapresse.ca/archives'
url_end = '.php'

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"><br>*** 3.2 Créer une liste vide***</span></div>

In [17]:
archives_url = []

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"><br>*** 3.3 Peupler la liste à partir d'une boucle *** (Imputer les nouvelles addresses. Partir de notre souche et intégrer les dates au four et à mesure que nous itérons dans la liste.) </span></div> 

In [18]:
for x in range(len(array_of_time)):   # Pour tout les éléments de la liste, 
    day = array_of_time[x].day        # capturer les données day, month, year
    month = array_of_time[x].month    
    year = array_of_time[x].year
    archives_url.append('{}/{}/{}/{}{}'.format(root_url,year,month,day,url_end)) # Stocker l'URL dans la nouvelle liste

In [23]:
archives_url[0:5] # Imprimer les 5 premiers liens

['http://www.lapresse.ca/archives/2012/12/1.php',
 'http://www.lapresse.ca/archives/2012/12/2.php',
 'http://www.lapresse.ca/archives/2012/12/3.php',
 'http://www.lapresse.ca/archives/2012/12/4.php',
 'http://www.lapresse.ca/archives/2012/12/5.php']

In [26]:
archives_url[-5:] # Imprimer les 5 derniers liens

['http://www.lapresse.ca/archives/2013/6/26.php',
 'http://www.lapresse.ca/archives/2013/6/27.php',
 'http://www.lapresse.ca/archives/2013/6/28.php',
 'http://www.lapresse.ca/archives/2013/6/29.php',
 'http://www.lapresse.ca/archives/2013/6/30.php']

In [28]:
print(len(archives_url)) # Afficher la longueur de la liste

212


<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Générer une telle liste nécessite d'examiner la structure du site web en question, de comprendre sa logique, de retenir les constantes (le début du site et le type '.php, .html') et générer les extensions souhaitées par la suite. Dans ce cas, il s'agit de l'année, du mois et du jour.

###### <span style="color: black; font-family: Babas Neue; font-size: 2em;">4. Explorer les liens des archives</span>

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Nous devons interroger la liste de liens des **archives** dans le but d'extraire les liens vers les **articles**. Chaque lien des archives est donc à la fois la cible d'une requête et le conteneur de d'autres liens, qui feront l'objet d'une requête (pour accéder au contenu de chaque article). <br><br><br>***4.1 Interroger un site (nous allons prendre le lien du 11 janvier 2013)***</span></div>

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"><br>*** 4.1.1 Envoyer une requête au serveur avec urllib.request***</span></div>

In [29]:
print (archives_url[41]) # Vérifier s'il s'agit du lien souhaité

http://www.lapresse.ca/archives/2013/1/11.php


In [30]:
request = urllib.request.Request(url = archives_url[41], data=None, headers=
                             {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_3) \
                             AppleWebKit/537.36 (KHTML, like Gecko) Chrome/35.0.1916.47 Safari/537.36'})

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"><br>*** 4.1.2 Ouvrir un objet BeautifulSoup4 avec la réponse de la requête.*** <br><br>Dit autrement, nous passons d'envoyer une demande d'information à recevoir une réponse. bs4 va nous aider dans la deuxième phase et va afficher le contenu de la réponse.</span></div>

In [31]:
source_code = bs4.BeautifulSoup(urllib.request.urlopen(request).read(),'lxml')

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">***<br><br>4.1.2 Utiliser la méthode prettify pour afficher le contenu. Il s'agit du code source de la page.*** (Désactivé pour sauver de l'espace. Enlever le # pour réactiver.) </span></div>

In [33]:
# print(source_code.prettify) 

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> L'intérêt est d'extraire un seul site pour commencer et de voir son contenu afin essayer d'en déduire la structure interne commune. Si vous observez bien, vous verrez que la liste des articles est balisée avec un code 'ul class="square square-spread"', qui signifie *unordered list*. Cela est très significatif, car nous avons exactement droit ici à une telle liste. Probablement que cette balise est donc utilisée comme standard par La Presse. Nous tenterons de l'utiliser pour tout les autres liens.  

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">***<br>4.1.3 Convertir le contenu en chaîne de caractères et tronquer la sélection souhaitée.***</span></div>

In [35]:
str_from_code = str(source_code)
start_idx = str_from_code.find('<ul class="square square-spread">')
skimed_str = str_from_code[start_idx:str_from_code.find('</ul>',start_idx)]
print (skimed_str[0:500], '\n') # Imprimer les 500 premiers caractères
print (skimed_str[-500:]) # Imprimer les 500 premiers caractères

<ul class="square square-spread">
<li>(05:06) <a alt="C'est arrivé un 11 janvier" href="http://www.lapresse.ca/le-droit/actualites/ephemerides/201301/04/01-4608380-cest-arrive-un-11-janvier.php" title="C'est arrivé un 11 janvier">C'est arrivé un 11 janvier</a></li>
<li>(05:00) <a alt="La LNH s'éloigne de Québec" href="http://www.lapresse.ca/le-soleil/opinions/carrefour/201301/08/01-4609376-la-lnh-seloigne-de-quebec.php" title="La LNH s'éloigne de Québec">La LNH s'éloigne de Québec</a></li>
<li>( 

nspirant aux résidences Monchénou</a></li>
<li>(09:35) <a alt="Collecte de sang fructueuse" href="http://www.lapresse.ca/le-droit/actualites/vie-communautaire/201301/14/01-4611048-collecte-de-sang-fructueuse.php" title="Collecte de sang fructueuse">Collecte de sang fructueuse</a></li>
<li>(09:39) <a alt="Bourses d'excellence" href="http://www.lapresse.ca/le-droit/actualites/vie-communautaire/201301/14/01-4611053-bourses-dexcellence.php" title="Bourses d'excellence">Bourses d'excellence</a></l

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> À travers ce codage, nous pouvons observer que les liens qui nous intéressent suivent du mot-clé 'href=' et termine toujours par l'extension '.php'. À partir de là, il ne reste que quelques manipulations pour accéder à notre dernière liste.

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">***<br>4.1.4 Diviser le texte par le mot-clé 'href' et lister les éléments. Enlever le premier index car il ne s'agit pas d'un site.***</span></div>

In [38]:
php_list = skimed_str.split('href="')[1:]
php_list[0:2]# Imprimer les deux premiers lien

['http://www.lapresse.ca/le-droit/actualites/ephemerides/201301/04/01-4608380-cest-arrive-un-11-janvier.php" title="C\'est arrivé un 11 janvier">C\'est arrivé un 11 janvier</a></li>\n<li>(05:00) <a alt="La LNH s\'éloigne de Québec" ',
 'http://www.lapresse.ca/le-soleil/opinions/carrefour/201301/08/01-4609376-la-lnh-seloigne-de-quebec.php" title="La LNH s\'éloigne de Québec">La LNH s\'éloigne de Québec</a></li>\n<li>(07:02) <a alt="Seul devant la cour, un bon choix?" ']

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">***<br>4.1.5 Modifier les éléments de la liste pour qu'ils s'arrêtent à l'extension '.php' ***</span></div>

In [42]:
for x in range(len(php_list)):
    php_list[x] = php_list[x][:php_list[x].find('.php')+4]
print(php_list[0:10], "\n") # Imprimer les dix premiers liens
print (len(php_list)) # Combien d'articles publiés le 11 janvier 2013?

['http://www.lapresse.ca/le-droit/actualites/ephemerides/201301/04/01-4608380-cest-arrive-un-11-janvier.php', 'http://www.lapresse.ca/le-soleil/opinions/carrefour/201301/08/01-4609376-la-lnh-seloigne-de-quebec.php', 'http://www.lapresse.ca/la-tribune/archives/la-nouvelle/actualites/201301/09/01-4609791-seul-devant-la-cour-un-bon-choix.php', 'http://www.lapresse.ca/la-tribune/archives/la-nouvelle/scene-culturelle/201301/09/01-4609793-impair-un-collectif-inspire-par-lart-de-rue.php', 'http://www.lapresse.ca/la-tribune/archives/la-nouvelle/actualites/201301/09/01-4609795-la-consigne-sociale-en-plein-emergence.php', 'http://www.lapresse.ca/le-droit/actualites/chroniques/denis-gratton/201301/10/01-4609974-evangeline-gabriel-et-le-chat.php', 'http://www.lapresse.ca/le-soleil/opinions/carrefour/201301/10/01-4610077-nadeau-dubois-futur-lieutenant-gouverneur.php', 'http://www.lapresse.ca/debats/editoriaux/francois-cardinal/201301/10/01-4610192-les-racines-de-la-colere.php', 'http://www.lapresse

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"><br>*** 4.2 Application de l'extraction à tout les liens.***</span></div>

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Pour obtenir tout les articles écris par la La Presse du 1er décembre au 30 juin, il suffit de généraliser ce que l'on vient de faire aux autres liens des archives et stocker les nouveaux dans une nouvelle variable. Vous pouvez exécuter le code ci-dessous, mais je vais le désactiver par défaut, car, pour le bien du tutoriel, je ne traiterai que d'une seule journée, soit celle que je viens de faire.

In [43]:
#metalist = []
#for x in range(len(archives_url)):
#
#    request = urllib.request.Request(url = archives_url[x], data=None, headers=
#                             {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_3) \
#                             AppleWebKit/537.36 (KHTML, like Gecko) Chrome/35.0.1916.47 Safari/537.36'})
#
#    source_code = bs4.BeautifulSoup(urllib.request.urlopen(request).read(),'lxml')
#    str_from_code = str(source_code)
#    start_idx = str_from_code.find('<ul class="square square-spread">')
#    skimed_str = str_from_code[start_idx:str_from_code.find('</ul>',start_idx)]
#    php_list = skimed_str.split('href="')[1:]
#    
#    for y in range(len(php_list)):
#        php_list[y] = php_list[y][:php_list[y].find('.php')+4]
#
#    metalist.extend(php_list)
    
#metalist

###### <span style="color: black; font-family: Babas Neue; font-size: 2em;">5. Explorer les liens des articles</span>

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Notre liste de liens vers les articles ayant été générée, nous allons pouvoir maintenant extraire d'elles l'information qui nous intéresse. Souvent, notre but est de rechercher quelque chose de particulier, comme dans notre situation où nous recherchons les articles ayant la mention IDLE NO MORE dans le texte. Il ne s'agit pas d'une comparaison entre les articles faisant mention et ceux pour qui cela n'est pas le cas, mais bien d'une comparaison entre les articles faisant mention du mouvement. Ainsi, nous n'avons pas réellement besoin des autres articles. <br><br><br><div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">Le module newspaper est utile pour les raisons suivantes : <br><br>
- Il nous évite de passer par la technique du départ, à savoir envoyer une requête avec urllib, traiter le contenu avec bs4, puis convertir en chaîne de caractères pour finalement scinder le texte à partir de balises dans le code. Newspaper fait ses étapes en une seule commande.<br>
- Le module crée un objet de type Article, ce qui lui infère de nombreux attributs. C'est pour cela que l'on ne peut pas l'utiliser lors de la première étape.<br>
</span></div>
  <br><br><br>***5.1 Interroger un site (soit le premier lien du 11 janvier 2013)***</span></div>

In [46]:
article_url = php_list[0]
print (article_url)

http://www.lapresse.ca/le-droit/actualites/ephemerides/201301/04/01-4608380-cest-arrive-un-11-janvier.php


<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">***<br>5.1.1  Utiliser le module newspaper pour récupérer une site***</span></div>

In [47]:
article = newspaper.Article(article_url, language ='fr') # Créer un objet article avec un URL
article.download() # Télécharger
article.parse() # Traiter
print (type(article)) # Afficher le type

<class 'newspaper.article.Article'>


<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">***<br>5.1.2 Utiliser l'attribut 'text' de l'objet article***</span></div>

In [48]:
print(article.text)

2001 - Le premier ministre québécois Lucien Bouchard annonce qu'il renonce à la politique active.

2008 - Décès de Sir Edmund Hillary, 88 ans, alpiniste et explorateur néo-zélandais qui a conquis l'Everest en 1953, en compagnie du sherpa Tenzing Norgay.

1998 - À Lahore, au Pakistan, 22 musulmans chiites sont tués par balles lors d'un office.

1998 - Des bandes armées assassinent 103 civils au sud-est d'Alger.

1995 - Les hockeyeurs s'entendent avec les propriétaires, après trois mois de litige dans la LNH.

1994 - Au départ de Robert Bourassa, Daniel Johnson devient premier ministre du Québec.

1993 - Richard Branson obtient des dommages de British Airways, pour diffamation contre sa compagnie Virgin Airways.

1988 - Un tribunal de Chicago ordonne à la pétrolière Amoco de payer 85,2 millions $ US en dommages, pour le naufrage de l'Amoco Cadiz sur des récifs devant Ploudalmézeau, en Bretagne, en mars 1978, la pire marée noire de l'histoire en Europe.

1987 - Près du port de Bassorah, l

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">***<br>5.1.3 Nos mots sont-il contenu dans l'article? ***</span></div>

In [49]:
'IDLE NO MORE' in str(article.text).upper()

False

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;">***<br>5.1.4 Vérifier les autres attributs ***</span></div>

In [50]:
article.title

"C'est arrivé un 11 janvier"

In [51]:
article.meta_img

'http://images.lpcdn.ca/924x615/201209/21/592329-lucien-bouchard.jpg'

In [52]:
article.meta_data

defaultdict(dict,
            {'apple-itunes-app': 'app-id=451829003, app-argument=cyberpresse://article/4608380',
             'article': {'publisher': 'https://www.facebook.com/LaPresseFB'},
             'description': "2001 - Le premier ministre québécois Lucien Bouchard annonce qu'il renonce à la politique active.",
             'fb': {'admins': 100001148387127, 'app_id': 166995983353903},
             'keywords': 'Gatineau, Ottawa',
             'og': {'description': "2001 - Le premier ministre québécois Lucien Bouchard annonce qu'il renonce à la politique active.",
              'image': 'http://images.lpcdn.ca/924x615/201209/21/592329-lucien-bouchard.jpg',
              'locale': 'fr_CA',
              'site_name': 'La Presse',
              'title': "C'est arrivé un 11 janvier | Éphémérides",
              'type': 'article',
              'url': 'http://www.lapresse.ca/le-droit/ephemerides/201301/04/01-4608380-cest-arrive-un-11-janvier.php'},
             'twitter': {'card': 's

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> Notre dernière étape est de générer une boucle dans la liste de liens, consulter ceux-ci et vérifier si le contenu principal contient les mots IDLE NO MORE. Si tel est le cas, nous ajouterons les informations sur l'article dans une base de données. Dans l'autre cas, nous passerons au lien suivant. 

In [58]:
article_data_list = []

for x in range(len(php_list)):
    article_url = newspaper.Article(php_list[x], language ='fr')
    article_url.download()
    str_html = article.html
    article_url.parse()

    keyws = article_url.meta_keywords
    authors = article_url.authors
    
    if len(authors) == 0 :
        authors = None
        
    text = article_url.text
    if 'IDLE NO MORE' in str(text).upper():
        print ('Link {} of {} : True'.format(x+1, len(php_list)))
        metadata = article_url.meta_data
        title = article_url.title
        metades = article_url.meta_description

        data_array = {'date' : array_of_time[41], 'authors' : authors, 'title' : title, 'text' : text, 'meta_keywords' : keyws,
                 'meta_description' : metades, 'meta_data' :metadata, 'url' : php_list[x]}
        article_data_list.append(pd.DataFrame(index=['LP_{}{}{}-{}'.format(array_of_time[41].year,array_of_time[41].month,array_of_time[41].day,x)],
                                          columns = 'authors date meta_data meta_description meta_keywords text title url'.split(), data = pd.Series(data_array).values.reshape(1,8)))
    else:
        print ('Link {} of {} : False'.format(x+1, len(php_list)))

http://www.lapresse.ca/le-droit/actualites/ephemerides/201301/04/01-4608380-cest-arrive-un-11-janvier.php 1 of 289
No
http://www.lapresse.ca/le-soleil/opinions/carrefour/201301/08/01-4609376-la-lnh-seloigne-de-quebec.php 2 of 289
No
http://www.lapresse.ca/la-tribune/archives/la-nouvelle/actualites/201301/09/01-4609791-seul-devant-la-cour-un-bon-choix.php 3 of 289
No
http://www.lapresse.ca/la-tribune/archives/la-nouvelle/scene-culturelle/201301/09/01-4609793-impair-un-collectif-inspire-par-lart-de-rue.php 4 of 289
No
http://www.lapresse.ca/la-tribune/archives/la-nouvelle/actualites/201301/09/01-4609795-la-consigne-sociale-en-plein-emergence.php 5 of 289
No
http://www.lapresse.ca/le-droit/actualites/chroniques/denis-gratton/201301/10/01-4609974-evangeline-gabriel-et-le-chat.php 6 of 289
No
http://www.lapresse.ca/le-soleil/opinions/carrefour/201301/10/01-4610077-nadeau-dubois-futur-lieutenant-gouverneur.php 7 of 289
No
http://www.lapresse.ca/debats/editoriaux/francois-cardinal/201301/10/0

In [108]:
df = pd.concat(article_data_list,axis=0)
df

authors                 date  \
LP_2013111-7      None  2013-01-11 00:00:00   
LP_2013111-26     None  2013-01-11 00:00:00   
LP_2013111-45     None  2013-01-11 00:00:00   
LP_2013111-168    None  2013-01-11 00:00:00   
LP_2013111-195    None  2013-01-11 00:00:00   
LP_2013111-210    None  2013-01-11 00:00:00   
LP_2013111-232    None  2013-01-11 00:00:00   

                                                        meta_data  \
LP_2013111-7    {'description': 'La campagne de protestations ...   
LP_2013111-26   {'description': 'La Conférence régionale des é...   
LP_2013111-45   {'description': 'On dira que Stephen Harper n'...   
LP_2013111-168  {'description': 'Puisqu'il craint une escalade...   
LP_2013111-195  {'description': 'La réunion entre Stephen Harp...   
LP_2013111-210  {'description': 'Quelques centaines d'Autochto...   
LP_2013111-232  {'description': 'Une vingtaine de personnes on...   

                                                 meta_description  \
LP_2013111-7    La campagne de protestations Idle no more a un...   
LP_2013111-26   La Conférence régionale des élus (CRE) de la G...   
LP_2013111-45   On dira que Stephen Harper n'avait pas vraimen...   
LP_2013111-168  Puisqu'il craint une escalade des tensions au ...   
LP_2013111-195  La réunion entre Stephen Harper et des leaders...   
LP_2013111-210  Quelques centaines d'Autochtones et de sympath...   
LP_2013111-232  Une vingtaine de personnes ont bloqué vendredi...   

                                                    meta_keywords  \
LP_2013111-7                                  [François Cardinal]   
LP_2013111-26   [Québec - Le Soleil - Actualités - Internation...   
LP_2013111-45                            [Chroniqueur, politique]   
LP_2013111-168  [Parti Conservateur, Parti Libéral, Bloc Québé...   
LP_2013111-195  [Québec - Le Soleil - Actualités - Internation...   
LP_2013111-210                                        [La Presse]   
LP_2013111-232                            [Actualités nationales]   

                                                             text  \
LP_2013111-7    La campagne de protestations Idle no more a un...   
LP_2013111-26   (Pointe-à-la-Croix) La Conférence régionale de...   
LP_2013111-45   On dira que Stephen Harper n'avait pas vraimen...   
LP_2013111-168  (Ottawa) Puisqu'il craint une escalade des ten...   
LP_2013111-195  La réunion entre Stephen Harper et des leaders...   
LP_2013111-210  Quelques centaines d'Autochtones et de sympath...   
LP_2013111-232  Une vingtaine de personnes ont bloqué vendredi...   

                                                            title  \
LP_2013111-7                             Les racines de la colère   
LP_2013111-26   La CRE de la Gaspésie et des Îles appuie Idle ...   
LP_2013111-45         Un autre rendez-vous de la dernière chance?   
LP_2013111-168      Les Premières nations deviennent une priorité   
LP_2013111-195  La réunion entre Harper et les Autochtones se ...   
LP_2013111-210  Manifestation au soutien des Autochtones à Mon...   
LP_2013111-232  Des autochtones bloquent un chemin de fer en N...   

                                                              url  
LP_2013111-7    http://www.lapresse.ca/debats/editoriaux/franc...  
LP_2013111-26   http://www.lapresse.ca/le-soleil/actualites/so...  
LP_2013111-45   http://www.lapresse.ca/debats/chroniques/vince...  
LP_2013111-168  http://www.lapresse.ca/actualites/politique/po...  
LP_2013111-195  http://www.lapresse.ca/le-soleil/actualites/po...  
LP_2013111-210  http://www.lapresse.ca/actualites/montreal/201...  
LP_2013111-232  http://www.lapresse.ca/actualites/national/201...

###### <span style="color: black; font-family: Babas Neue; font-size: 2em;">6. Traitement et exportation</span>

<div style="text-align: justify"> <span style="color: black; font-family: Babas Neue; font-size: 1em;"> En guise de conclusion, nous allons effectuer quelques changements à notre base de données. Premièrement, le module newspaper - par sa faute ou par la différence de protocole chez La Presse -, n'a pas enregistrer aucune donnée sur les auteurs. Néanmoins, nous pouvons retirer cette information des méta-données, que nous avons stocké. Nous allons remplir la colonne des auteurs en se fiant aux méta-données. Deuxièment, les dates contiennent des informations sur les heures, qui non seulement ne sont pas pertinentes pour nous, mais sont également erronées (ce n'est pas à 00:00 que l'article a été publié). Nous les enlèverons. Enfin, nous allons enregistrer notre base de données, en format .csv que je privilégie, mais également en format .xls.

In [109]:
l = []
for x in range(len(df)):
    l.append(df['meta_data'][x]['og']['title'].split(' | ')[1].split(',')[0])
    df.ix[x,'date'] = str(df.ix[x,'date']).split()[0]
    
df['authors'] = l

In [110]:
df

authors        date  \
LP_2013111-7              François Cardinal  2013-01-11   
LP_2013111-26                  Gilles Gagné  2013-01-11   
LP_2013111-45              Vincent Marissal  2013-01-11   
LP_2013111-168        Joël-Denis Bellavance  2013-01-11   
LP_2013111-195  Lina Dib et Stéphanie Marin  2013-01-11   
LP_2013111-210   Philippe Teisceira-Lessard  2013-01-11   
LP_2013111-232                     National  2013-01-11   

                                                        meta_data  \
LP_2013111-7    {'description': 'La campagne de protestations ...   
LP_2013111-26   {'description': 'La Conférence régionale des é...   
LP_2013111-45   {'description': 'On dira que Stephen Harper n'...   
LP_2013111-168  {'description': 'Puisqu'il craint une escalade...   
LP_2013111-195  {'description': 'La réunion entre Stephen Harp...   
LP_2013111-210  {'description': 'Quelques centaines d'Autochto...   
LP_2013111-232  {'description': 'Une vingtaine de personnes on...   

                                                 meta_description  \
LP_2013111-7    La campagne de protestations Idle no more a un...   
LP_2013111-26   La Conférence régionale des élus (CRE) de la G...   
LP_2013111-45   On dira que Stephen Harper n'avait pas vraimen...   
LP_2013111-168  Puisqu'il craint une escalade des tensions au ...   
LP_2013111-195  La réunion entre Stephen Harper et des leaders...   
LP_2013111-210  Quelques centaines d'Autochtones et de sympath...   
LP_2013111-232  Une vingtaine de personnes ont bloqué vendredi...   

                                                    meta_keywords  \
LP_2013111-7                                  [François Cardinal]   
LP_2013111-26   [Québec - Le Soleil - Actualités - Internation...   
LP_2013111-45                            [Chroniqueur, politique]   
LP_2013111-168  [Parti Conservateur, Parti Libéral, Bloc Québé...   
LP_2013111-195  [Québec - Le Soleil - Actualités - Internation...   
LP_2013111-210                                        [La Presse]   
LP_2013111-232                            [Actualités nationales]   

                                                             text  \
LP_2013111-7    La campagne de protestations Idle no more a un...   
LP_2013111-26   (Pointe-à-la-Croix) La Conférence régionale de...   
LP_2013111-45   On dira que Stephen Harper n'avait pas vraimen...   
LP_2013111-168  (Ottawa) Puisqu'il craint une escalade des ten...   
LP_2013111-195  La réunion entre Stephen Harper et des leaders...   
LP_2013111-210  Quelques centaines d'Autochtones et de sympath...   
LP_2013111-232  Une vingtaine de personnes ont bloqué vendredi...   

                                                            title  \
LP_2013111-7                             Les racines de la colère   
LP_2013111-26   La CRE de la Gaspésie et des Îles appuie Idle ...   
LP_2013111-45         Un autre rendez-vous de la dernière chance?   
LP_2013111-168      Les Premières nations deviennent une priorité   
LP_2013111-195  La réunion entre Harper et les Autochtones se ...   
LP_2013111-210  Manifestation au soutien des Autochtones à Mon...   
LP_2013111-232  Des autochtones bloquent un chemin de fer en N...   

                                                              url  
LP_2013111-7    http://www.lapresse.ca/debats/editoriaux/franc...  
LP_2013111-26   http://www.lapresse.ca/le-soleil/actualites/so...  
LP_2013111-45   http://www.lapresse.ca/debats/chroniques/vince...  
LP_2013111-168  http://www.lapresse.ca/actualites/politique/po...  
LP_2013111-195  http://www.lapresse.ca/le-soleil/actualites/po...  
LP_2013111-210  http://www.lapresse.ca/actualites/montreal/201...  
LP_2013111-232  http://www.lapresse.ca/actualites/national/201...

In [113]:
df.to_excel('lapresse_11janv.xls')
df.to_csv('lapresse_11janv.csv')